clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/PersistenceDAO.scala

```diff
@@ -78,6 +100,84 @@ trait PersistenceDAO extends LazyLogging {
       )
       logger.debug(s"Wrote document $document to ${written.toUri}")
     }
+
+  /**
+    * Throw or return the `ClioDocument` at `path`.
+    * @param path to a JSON-encoded `ClioDocument``
+    * @tparam D is the type of `ClioDocument` at `path`
+    * @return the `ClioDocument` at `path`
+    */
+  def documentFromPath[D <: ClioDocument: Decoder](path: Path): D = {
+    decode[D](new String(Files.readAllBytes(path))).fold(throw _, identity)
+  }
+
+  /**
+    * @param path in some params `FileSystem`
+    * @return true if `path` is a "params file"
+    */
+  def isParamsFile(path: Path): Boolean = {
+    val s = path.toString
+    val parts = s.split(File.separator)
+    val last = parts.last
+    val isJson = last.endsWith(".json")
+    val hasUuid = last.split('-').length == 5
+    parts.length == 6 && isJson && hasUuid
+  }
+
+  /**
+    * Return the `ClioDocument`s under `rootdir` upserted no earlier
+    * than `withId`.
+    *
+    * @param rootDir is path to a bucket of "params files"
+    * @param withId is the path of earliest "params file" to return
+    * @param ec is an `ExecutionContext`
+    * @tparam D is the type of `ClioDocument` expected under `rootDir`
+    * @return the `ClioDocument` at `withId` and all later ones
+    */
+  def sinceId[D <: ClioDocument: Decoder](rootDir: Path, withId: Path)(
+    implicit ec: ExecutionContext
+  ): Future[Seq[D]] = {
+    val before = (p: Path) => p.compareTo(withId) < 0
+    val after = Directory
+      .walk(rootDir)
+      .filter(p => isParamsFile(p))
+      .filter(!before(_))
+      .runWith(Sink.seq)
+      .map(_.sortBy(before))
+    Source
+      .fromFuture(after)
+      .mapConcat(identity)
+      .via(Flow[Path].map(p => documentFromPath(p)))
+      .runWith(Sink.seq[D])
+  }
+
+  /**
+    * Return the GCS bucket metadata files for `index` upserted on or
+    * after `upsertId`.  Decode the files from JSON as `ClioDocument`s
+    * sorted by `upsertId`.
+    *
+    * @param upsertId of the last known upserted document
+    * @param index  to query against.
+    * @param ec is the implicit ExecutionContext
+    * @tparam D is a ClioDocument type with a JSON Decoder
+    * @return a Future Seq of ClioDocument
+    */
+  def getAllSince[D <: ClioDocument: Decoder](
+    upsertId: UUID,
+    index: ElasticsearchIndex[D]
+  )(implicit ec: ExecutionContext): Future[Seq[D]] = {
+    val rootDir = rootPath.resolve(index.rootDir)
+    val suffix = s"/${upsertId}.json"
+    Directory
+      .walk(rootDir)
+      .filter(p => isParamsFile(p))
```

Comment:
Do we need this filter?

Reply:
Nope. At least not when running against jimfs.